### PR TITLE
Improve trx_generator http client

### PR DESCRIPTION
**Pull request by heifner**
When error conditions (eof) have the `http_plugin` `shutdown_both` on socket. Also call `close()` for good measure. I actually tried this explicit shutdown in the destructor but it seemed to make no difference.

Modify the `trx_generator` to better handle errors from client and also specify no keep alive.

This improved performance as can be seen by comparing:
This branch: https://github.com/AntelopeIO/leap/actions/runs/6356928532  15001 TPS
With main: https://github.com/AntelopeIO/leap/actions/runs/6356929999  14001 TPS

### CODE REVIEW

File: tests/trx_generator/http_client_async.hpp [hunk 1]
```diff
@@ -73,7 +73,9 @@ class session : public std::enable_shared_from_this<session> {
       req_.set(http::field::host, host);
       req_.set(http::field::user_agent, BOOST_BEAST_VERSION_STRING);
       req_.set(http::field::content_type, content_type);
+      req_.set(http::field::connection, "close");
       req_.body() = std::move(request_body);
+      req_.keep_alive(false);
```
Comment (greg7mdp) [inline tests/trx_generator/http_client_async.hpp:78]: Why these changes?

File: plugins/http_plugin/include/eosio/http_plugin/beast_http_session.hpp [hunk 2]
```diff
@@ -503,7 +503,8 @@ class beast_http_session : public detail::abstract_conn,
       try {
          // Send a shutdown signal
          beast::error_code ec;
-         socket_.shutdown(Socket::shutdown_send, ec);
+         socket_.shutdown(Socket::shutdown_both, ec);
+         socket_.close(ec);
```
[the remark this answers is not in the document]
Comment (heifner) [reply plugins/http_plugin/include/eosio/http_plugin/beast_http_session.hpp:507]: This change less about performance and more about leaving a socket in timed wait on client abandoning a socket. The eof is used for error conditions, so makes sense to shutdown both.

File: tests/trx_generator/http_client_async.hpp (hunk 1 above)
Comment (heifner) [reply tests/trx_generator/http_client_async.hpp:78]: It would if the trx_generator reused a socket. It creates a new connection for each request. We really need to fix that because on my machine I run out of ports.

Comment (greg7mdp) [reply tests/trx_generator/http_client_async.hpp:78]: Thanks. For posterity, maybe add a comment that we disable `keep_alive` because the `trx_generator ` doesn't reuse sockets, but this needs to be changed, and then `keep_alive` should be re-enabled.

File: plugins/http_plugin/include/eosio/http_plugin/beast_http_session.hpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (spoonincode) [reply plugins/http_plugin/include/eosio/http_plugin/beast_http_session.hpp:507]: This change got rid of `TIME_WAIT` states?

[the remark this answers is not in the document]
Comment (heifner) [reply plugins/http_plugin/include/eosio/http_plugin/beast_http_session.hpp:507]: No, the keep_alive change helped with the `TIME_WAIT`. This was for when it was already in a bad state.

**Review by greg7mdp — APPROVED** [review]

File: plugins/http_plugin/include/eosio/http_plugin/beast_http_session.hpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (spoonincode) [reply plugins/http_plugin/include/eosio/http_plugin/beast_http_session.hpp:507]: Not sure I understand what the purpose of this entire `do_eof()` function is. Letting the socket's shared_ptr fall out of scope is going to close it anyways.

[the remark this answers is not in the document]
Comment (heifner) [reply plugins/http_plugin/include/eosio/http_plugin/beast_http_session.hpp:507]: I'll try experimenting with just removing it.

[the remark this answers is not in the document]
Comment (heifner) [reply plugins/http_plugin/include/eosio/http_plugin/beast_http_session.hpp]: In https://github.com/AntelopeIO/leap/pull/1699/commits/4f4f5d3350723758b25554be37fb8e6d57f50b12 removed `do_eof`, resulted in `read-only-trx-parallel-test` failing do to `Failure when receiving data from the peer` which looks like it is http error 56. This test does a lot of http requests to nodeos. 

It does look like the `do_eof` is needed. The example boost code also has the `do_eof`.
@spoonincode I will revert https://github.com/AntelopeIO/leap/pull/1699/commits/4f4f5d3350723758b25554be37fb8e6d57f50b12 unless you can find something wrong in my removal of it.

[the remark this answers is not in the document]
Comment (heifner) [reply plugins/http_plugin/include/eosio/http_plugin/beast_http_session.hpp]: However, with the removal of `do_eof`, this branch reported a `15501` TPS.

[the remark this answers is not in the document]
Comment (spoonincode) [reply plugins/http_plugin/include/eosio/http_plugin/beast_http_session.hpp]: Seems like we might have a problem somewhere; I'm still not understanding why it's needed. But yeah might as well put it back for now.

**Review by spoonincode — APPROVED** [review]
